### DISCUSSION TITLE
The startup of our project

### DISCUSSION THREAD

**Discussion by Peng-Rao** [AMSC-24-25/03-rsvd-team-3#3] [Ideas]
I want to share my idea of a startup of our project. According to the scoring rules mentioned, programming techniques beyond the course content and Programming level related to the course content account for a large proportion, so I think we should focus on the following aspects:

- `C++20`: I think we should use the C++20 introduced in the course content.
- `Object-oriented`: We should use object-oriented programming to design the project.
- `Parrallelism`: Prof has introduced the parallelism in the course content, personally, I think MPI + OpenMP is a good choice, also, Eigen has implemente parallel algorithms, but I am not very familiar with them, we can also do some optimization based on his parallel algorithm.
- `Optimization`: Profiling is a good way to optimize our code, we can use `gprof` or `valgrind` to profile our code, and optimize our code based on the profiling results.

To better organize our project, I think we should use the following tools:

- `CMake+Vcpkg`: We can use CMake to manage our project. And Vcpkg(like python pip) is a good package manager for C++. We can use it to manage our dependencies. This way we can develop on our own platform without version differences. Microsoft guide is here: [Vcpkg](https://learn.microsoft.com/en-us/vcpkg/get_started/get-started?pivots=shell-bash)
- `.clang-tidy` and `.clang-format`: Use them to ensure the code quality and style. Actualy, they are included in many IDEs, like VSCode, CLion, etc.
- `test` and `benchmark`: I saw that the professor has uploaded relevant information(`Catch2` or `gtest`), such libraries are easy to install using Vcpkg, we can use them to test our code and optimize our code.

If you have any other ideas, please let me know. I am looking forward to your reply. Thank you!

**Comment by jlclaudio** [comment]
Why not just use the hpc_env that set in the class, instead of setting another environment which cost much time.

**Reply by Peng-Rao** [reply]
The hpc_env environment is not complete and still requires the installation of third-party libraries such as unit tests and benchmarks. Of course, you can also install them in the hpc_env environment, I just want the project structure to be neat.

**Comment by jlclaudio** [comment]
and I think we will not use the MPI which is used in the distributed system. the openMP is enough

**Reply by Peng-Rao** [reply]
MPI can also be used with multi-core processors
